### DISCUSSION TITLE
How to name queries?

### DISCUSSION THREAD

**Discussion by mmkal** [gajus/slonik#288] [Ideas]
Hey @gajus - an idea I've been mulling over that I think could fit nicely into slonik is an ability to "name" queries. Honeycomb have written a good blog about some of the advantages: [Show your query you love it by naming it](https://www.honeycomb.io/blog/naming-queries-made-better/). It enables recording performance metrics for a given query easily, even if the query itself changes over time - which would complicate any automatic query id generation techniques.

The change in slonik would be very small:

```diff
export type SqlSqlTokenType = {
  readonly sql: string,
  readonly type: typeof tokens.SqlToken,
  readonly values: readonly PrimitiveValueExpressionType[],
+ readonly name?: string
};
```

Then the `name` property could be passed to `executeQuery` and added to `originalQuery` so it reaches the various interceptors:

```diff
  const originalQuery = {
    sql: rawSql,
    values,
+   name,
  };
```

From there end-users can do what they like with it, including logging timing/success metrics, etc.

```ts
afterQueryExecution: async (query) => {
  console.log({query: query.originalQuery.name, requestId: ...})
  return null
}
```

To use it, there could be a helper function:

```ts
export const usersByNameQuery = (name: string) => sql.named('UsersByName')`
  select * from usr where name = ${name}
`

export const usersByEmailQuery = (email: string) => sql.named('UsersByEmail')`
  select * from usr where email = ${email}
`

const users = await pool.any(usersByNameQuery('Bob'))
```

With that in place, it may be possible to do some clever stuff using [@slonik/typegen](https://www.npmjs.com/package/@slonik/typegen) to infer the correct type based on the name.

I'd be happy to open a PR along those lines if you'd be open to it?

**Comment by gajus** [comment]
How would someone construct a query?

**Reply by mmkal** [reply]
See the example - something like

```ts
const usersByEmail = sql.named('UsersByEmail')`
  select * from usr where email = ${email}
`

const users = await pool.query(usersByEmail)
```
